pkg/kubelet/preemption/preemption.go

```diff
@@ -0,0 +1,282 @@
+	// if no error is returned, preemption succeeded and the pod is safe to admit.
+	return err == nil, nil, err
+}
+
+// freeRequests takes a list of insufficient resources, and attempts to free them by evicting pods
+// based on requests.  For example, if the only insufficient resource is 200Mb of memory, this function could
+// evict a pod with request=250Mb.
+func (c *CriticalPodAdmissionHandler) evictPodsToFreeRequests(insufficientResources admissionRequirementList) error {
+	nonCriticalPods := filterPods(c.getPodsFunc(), func(p *v1.Pod) bool { return !kubetypes.IsCriticalPod(p) })
+	podsToPreempt, err := getPodsToPreempt(nonCriticalPods, insufficientResources)
+	if err != nil {
+		return fmt.Errorf("preemption: error finding a set of pods to preempt: %v", err)
+	}
+	glog.Infof("preemption: attempting to evict pods %v, in order to free up resources: %s", podsToPreempt, insufficientResources.toString())
+	for _, pod := range podsToPreempt {
+		status := v1.PodStatus{
+			Phase:   v1.PodFailed,
+			Message: message,
+			Reason:  reason,
+		}
+		// record that we are evicting the pod
+		c.recorder.Eventf(pod, v1.EventTypeWarning, reason, message)
+		// this is a blocking call and should only return when the pod and its containers are killed.
+		err := c.killPodFunc(pod, status, nil)
+		if err != nil {
+			return fmt.Errorf("preemption: pod %s failed to evict %v", format.Pod(pod), err)
+		}
+		glog.Infof("preemption: pod %s evicted successfully", format.Pod(pod))
+	}
+	return nil
+}
+
+// getPodsToPreempt returns a list of pods that could be preempted to free requests >= requirements
+func getPodsToPreempt(pods []*v1.Pod, requirements admissionRequirementList) ([]*v1.Pod, error) {
+	bestEffortPods := filterPods(pods, func(p *v1.Pod) bool { return qos.GetPodQOS(p) == v1.PodQOSBestEffort })
+	burstablePods := filterPods(pods, func(p *v1.Pod) bool { return qos.GetPodQOS(p) == v1.PodQOSBurstable })
+	guaranteedPods := filterPods(pods, func(p *v1.Pod) bool { return qos.GetPodQOS(p) == v1.PodQOSGuaranteed })
+
+	// We only evict guaranteed pods if we "have to".  This implies that guaranteed pods only need to fulfil requirements
+	// which are not fulfilled by all other burstable or best effort pods.
+	guarateedToEvict, err := getPodsToPreemptByDistance(guaranteedPods, requirements.subtract(append(bestEffortPods, burstablePods...)...))
+	if err != nil {
+		return nil, err
+	}
+	// Once we know the guaranteed pods being evicted, we can subtract their requirements from what we need to meet with burstable pods
+	// In similar fashion to guaranteed pods, we only need to consider requirements not met by all best effort pods.
+	burstableToEvict, err := getPodsToPreemptByDistance(burstablePods, requirements.subtract(append(bestEffortPods, guarateedToEvict...)...))
+	if err != nil {
+		return nil, err
+	}
+	bestEffortToEvict, err := getPodsToPreemptByDistance(bestEffortPods, requirements.subtract(append(burstableToEvict, guarateedToEvict...)...))
```

Comment:
There is no point in evaluating besteffort since they will not include any resource specifications.